✨ Add Rules Frontmatter Validator

Aibono1225 opened the pull request:
Relates to https://github.com/SSWConsulting/SSW.Rules/issues/1107

Changes made:
- Added GitHub action to check frontmatter in Markdown files on PRs to the main branch
- Left a comment on the PR if there is invalid frontmatter in `rule.md`
- Added JSON schema to validate frontmatter data


![image](https://github.com/SSWConsulting/SSW.Rules.Content/assets/127192800/909ed264-af75-4083-a9be-75e12430e310)
**Figure: GitHub comments to notice incorrect frontmatter**

![image](https://github.com/SSWConsulting/SSW.Rules.Content/assets/127192800/60b6e910-0cae-4d96-95b8-7737b1c5446c)
**Figure: Another comment example**

<!---
Thanks for contributing to SSW.Rules!

Have you seen our Rules to Better Pull Requests?
https://www.ssw.com.au/rules/rules-to-better-pull-requests/

Please provide a good title and a short description of your pull request
See https://www.ssw.com.au/rules/write-a-good-pull-request/ for further guidance

Did you do an over the shoulder review?
https://www.ssw.com.au/rules/over-the-shoulder-prs
-->

bradystroud commented on scripts/frontmatter-validator/frontmatter-validator.js:
```diff
@@ -0,0 +1,57 @@
+  if (!isValid) {
+    console.log(`Invalid Frontmatter detected in ${filePath.replaceAll('../', '')}, see details:`);
+    validate.errors.forEach((item) => {
+      if (item.keyword == 'errorMessage' || item.keyword == 'required') {
```
```suggestion
      if (item.keyword === 'errorMessage' || item.keyword === 'required') {
```
Use === to be consistent

Aibono1225 replied:
Done ✅

bradystroud commented on scripts/frontmatter-validator/frontmatter-validator.js:
```diff
@@ -0,0 +1,57 @@
+}
+
+function parseFrontmatter(fileContents) {
+  const frontmatterMatch = /^---([\s\S]*?)---/.exec(fileContents);
+  const frontmatterString = frontmatterMatch[1];
+  const frontmatter = yaml.load(frontmatterString, {
+    schema: yaml.JSON_SCHEMA,
+  });
+  return frontmatter;
```
I think there should be some error handling here... what if their is no frontmatter?
What if the `---` is missing?

Aibono1225 replied:
Done - Error handling has been added. Please see my comment below. 🙂

bradystroud commented on scripts/frontmatter-validator/frontmatter-validator.js:
```diff
@@ -0,0 +1,57 @@
+    console.log(`Invalid Frontmatter detected in ${filePath.replaceAll('../', '')}, see details:`);
+    validate.errors.forEach((item) => {
+      if (item.keyword == 'errorMessage' || item.keyword == 'required') {
+        console.log(`- ${item.message}`)
```
missing `;`
```suggestion
        console.log(`- ${item.message}`);
```

Aibono1225 replied:
Done ✅

bradystroud commented on .github/workflows/validate-frontmatter.yml:
Remove the useless comments, and maybe add some comments explaining each step

bradystroud reviewed (changes requested):
This is looking really good 😊
A few changes, but overall pretty awesome!

Aibono1225 commented on scripts/frontmatter-validator/schema.json:
```diff
@@ -0,0 +1,69 @@
+      }
+    }
+  },
+  "required": ["type", "title", "guid", "uri", "created", "authors"],
```
Unfortunately, in JSON Schema Draft 4+, the `required` keyword must be set as an array as per the specification. 😣
See https://json-schema.org/draft/2020-12/json-schema-validation.html#name-required.

Aibono1225 commented:
Hi @bradystroud 😊,

New changes have been made based on your feedback:
- Added schema for category and top-category
- Added error handling for parsing frontmatter

![image](https://github.com/SSWConsulting/SSW.Rules.Content/assets/127192800/2e81ddfb-1d8c-43e7-aeba-d290a1a9611c)
**Figure: I tested the workflow on my fork, and here are a few comments I got from it**

Aibono1225 commented on scripts/frontmatter-validator/frontmatter-validator.js:
```diff
@@ -0,0 +1,57 @@
+  const eventType = process.env.GITHUB_EVENT_NAME;
+
+  if (eventType === 'pull_request') {
+    if (process.argv[2] && process.argv[2].length > 0) {
```
Done ✅

bradystroud commented on scripts/frontmatter-validator/frontmatter-validator.js:
```diff
@@ -0,0 +1,93 @@
+    validator.addSchema(schemas[key], key);
+  }
+
+  return validator
```
```suggestion
  return validator;
```

Aibono1225 replied:
Done 🙂

bradystroud commented on scripts/frontmatter-validator/rule-schema.json:
Can you move all these to a `/schema` folder?

Aibono1225 replied:
✅ Done

bradystroud commented on scripts/frontmatter-validator/frontmatter-validator.js:
```diff
@@ -0,0 +1,93 @@
+
+  if (eventType === 'pull_request') {
+
+    // process.argv[2] represents the argument provided via command-line input when running the script.
+    // In the code below, it is a comma-separated list of changed files.
+    if (process.argv[2] && process.argv[2].length > 0) {
```
I think both comments can be replaced with a single line of code

```js
const filesChanged  = process.argv[2];
```

Then the variable name explains what it is.
What do you think?

Aibono1225 replied:
Done - Love this idea

bradystroud commented on scripts/frontmatter-validator/frontmatter-validator.js:
```diff
@@ -0,0 +1,93 @@
+}
+
+function validateFrontmatter(filePath) {
+  const fileContents = fs.readFileSync(filePath, 'utf8');
```
I think this could use some error handling

What if a file is deleted? Would GitHub still count it as a changed file? If it does this code will try read it

Aibono1225 replied:
Done - error handling has been added. Now the code will check if it exists the file before reading it. 🙂

bradystroud reviewed (approved):
LGTM 🙂
